docs/sources/reference/api/docker_remote_api_v1.17.md

```diff
@@ -802,6 +804,138 @@ Status Codes:
 -   **404** – no such container
 -   **500** – server error
 
+### Get files or folders from a container as a Tar Archive.
+
+`GET /containers/(id)/copy`
+
+Get an archive of a resource in the filesystem of container `id`.
+
+Query Parameters:
+
+-   **path** - The resource in the container's filesystem. If not an absolute
+        path, it is relative to the container's working directory. If path is
+        inside the mount point of a volume, the volume's data is archived. If
+        the path specifies a directory and a volume is mounted lower in that
+        directory hierarchy, the volume's contents are not included in the
+        archive. Required.
+
+**Example request**:
+
+        GET /containers/8cce319429b2/copy?path=/foo/ HTTP/1.1
+
+**Example response**:
+
+        HTTP/1.1 200 OK
+        Content-Type: application/x-tar
+
+        {{ TAR STREAM }}
+
+Status Codes:
+
+-   **200** – no error
+-   **400** - client error, details in JSON response body, one of:
+        - path cannot be empty (missing path query parameter)
+        - no such file or directory (path does not exist)
+        - not a directory (path ends with a separator but exists as a file)
+-   **404** – no such container
+-   **500** – server error
+
+### Upload files or folders to a container from a Tar Archive.
+
+`PUT /containers/(id)/copy`
+
+Upload an archive of a resource to be extracted to a path in the filesystem
+of container `id`.
+
+Query Parameters:
+
+-   **srcPath** - Specifies the original path from which the payload archive
```

Comment:
You're right - you can't tell from just the basename. It doesn't require it to end in a path separator. The way it works is that if you copy a file `cp /foo/bar /destDir` then the archive will have only that file and the Tar header for that entry will have `Name: bar` and *wont* have `TypeFlag == TypeDir`. If `bar` was a directory then that archive would have a first entry with `Name: bar` still, yes, but would have `TypeDir`. If the copy was specified with a trailing separator as in `cp /foo/bar/ /destDir` then `archive.CopyTo` will treat it as directory contents and not check the first archive entry.

> If its optional, and you can still get the job done, I would remove it.

Yup, I'm updating the PR to make it optional now.